cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -514,4 +514,67 @@ public static DatasetModule from(String namespaceId, String moduleId) {
       return new DatasetModule(Namespace.from(namespaceId), moduleId);
     }
   }
+
+  /**
+   * Dataset Instance Id identifies a given dataset instance.
+   */
+  public static final class DatasetInstance {
+    private final Namespace namespace;
+    private final String instanceId;
+
+    public DatasetInstance(final Namespace namespace, final String instanceId) {
+      Preconditions.checkNotNull(namespace, "Namespace cannot be null.");
+      Preconditions.checkNotNull(instanceId, "Dataset instance id cannot be null.");
+      Preconditions.checkArgument(isValidDatasetId(instanceId), "Invalid characters found in dataset instance id. '" +
+        instanceId + "'. Instance id can contain alphabets, numbers or _, -, . or $ characters");
+      this.namespace = namespace;
+      this.instanceId = instanceId;
+    }
+
+    public Namespace getNamespace() {
+      return namespace;
+    }
+
+    public String getNamespaceId() {
+      return namespace.getId();
+    }
+
+    public String getId() {
+      return instanceId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      DatasetInstance that = (DatasetInstance) o;
+      return namespace.equals(that.namespace) && instanceId.equals(that.instanceId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(namespace, instanceId);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+        .add("namespace", namespace)
```

Comment:
Great catch! Infact, I hit that problem in an upcoming PR. Going to override ``toString()`` in ``Id.Namespace`` to return just the ``String`` ``namespaceId`` that it wraps.